File: base/src/main/vsdk/toolkit/environment/geometry/surface/polygon/Polygon2D.java
package vsdk.toolkit.environment.geometry.surface.polygon;

import java.io.Serial;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import vsdk.toolkit.common.Vertex2D;
import vsdk.toolkit.common.Ray;
import vsdk.toolkit.common.dataStructures.BinaryTreeNode;
import vsdk.toolkit.environment.geometry.RayHit;
import vsdk.toolkit.environment.geometry.surface.Surface;
import vsdk.toolkit.processing.polygonClipper.PolygonProcessor;

public class Polygon2D extends Surface
{
    @Serial
    private static final long serialVersionUID = 20260419L;

    public List<_Polygon2DContour> loops;
    private _Polygon2DContour currentLoop;
    // headNode is the head node of the n-ary tree codified in a binary tree
    // in left child-right sibling way.
    private BinaryTreeNode<_Polygon2DContour> headNode;
    
    public Polygon2D()
    {
        loops = new ArrayList<_Polygon2DContour>();
        nextLoop();
    }

    public void addVertex(double x, double y, double r, double g, double b)
    {
        currentLoop.addVertex(x, y, r, g, b);
    }

    public void addVertex(double x, double y)
    {
        currentLoop.addVertex(x, y);
    }

    public void pushVertex(double x, double y)
    {
        currentLoop.pushVertex(x, y);
    }

    public final void nextLoop()
    {
        currentLoop = new _Polygon2DContour();
        loops.add(currentLoop);
    }

    public final void eraseLastLoop()
    {
        int i;
        _Polygon2DContour loop;
        
        if(loops.size() > 1) {
            // Remove also the holes of that loop. LRR.
            for(i=0; i<loops.size(); ++i) {
                loop = loops.get(i);
                if(loop.getExteriorContour() == loops.get(loops.size()-1)) {
                    loops.remove(i);
                    --i;
                }
            }
            if(loops.size() > 1) {
                loops.remove(loops.size()-1);
                currentLoop = loops.get(loops.size()-1);
            }
            // Rebuild the n-ary tree codified in a binary tree.
            PolygonProcessor.classifyContourHoles(this);
        }
    }
    
    public void invert()
    {
        for(_Polygon2DContour p2DContour : loops) {
            Collections.reverse(p2DContour.vertices);
        }
    }
    
    /**
    @return a new 6 valued double array containing the coordinates of a min-max
    bounding box for current geometry.
    */
    @Override
    public double[] getMinMax()
    {
        double minMax[];
        _Polygon2DContour l;
        Vertex2D v;
        int i;
        int j;
        minMax = new double[6];

        double minX = Double.MAX_VALUE;
        double minY = Double.MAX_VALUE;
        double maxX = -Double.MAX_VALUE;
        double maxY = -Double.MAX_VALUE;

        for ( i = 0; i < loops.size(); i++ ) {
            for ( j = 0; j < loops.get(i).vertices.size() - 1; j++ ) {
                v = loops.get(i).vertices.get(j);
                if ( v.x > maxX ) {
                    maxX = v.x;
                }
                if ( v.x < minX ) {
                    minX = v.x;
                }
                if ( v.y > maxY ) {
                    maxY = v.y;
                }
                if ( v.y < minY ) {
                    minY = v.y;
                }
            }
        }

        minMax[0] = minX;
        minMax[1] = minY;
        minMax[2] = 0;
        minMax[3] = maxX;
        minMax[4] = maxY;
        minMax[5] = 0;

        return minMax;
    }

    public Ray doIntersection(Ray inOut_ray)
    {
        return null;
    }

    @Override
    public boolean doIntersection(Ray inRay, RayHit outHit)
    {
        return false;
    }
    public void
    doExtraInformation(Ray inRay, double intT, 
                       RayHit outData)
    {

    }
    
    /**
     * @return the headNode
     */
    public BinaryTreeNode<_Polygon2DContour> getHeadNode() {
        return headNode;
    }

    /**
     * @param headNode the headNode to set
     */
    public void setHeadNode(BinaryTreeNode<_Polygon2DContour> headNode) {
        this.headNode = headNode;
    }
}

File: cpp/base/src/main/vsdk/toolkit/environment/geometry/surface/polygon/Polygon2D.cpp
#include "vsdk/toolkit/environment/geometry/surface/polygon/Polygon2D.h"
#include "vsdk/toolkit/environment/geometry/surface/polygon/_Polygon2DContour.h"
#include "vsdk/toolkit/environment/geometry/elements/Vertex2D.h"
#include "vsdk/toolkit/environment/geometry/elements/Ray.h"
#include "vsdk/toolkit/environment/geometry/elements/RayHit.h"
#include "vsdk/toolkit/common/dataStructures/BinaryTreeNode.h"
#include "vsdk/toolkit/common/VSDK.h"
#include "java/util/ArrayList.txx"

Polygon2D::Polygon2D() : currentLoop(nullptr), headNode(nullptr) { nextLoop(); }

Polygon2D::~Polygon2D()
{
    for (long int i = 0; i < loops.size(); i++) delete loops[i];
}

void Polygon2D::addVertex(double x, double y, double r, double g, double b) { currentLoop->addVertex(x,y,r,g,b); }
void Polygon2D::addVertex(double x, double y) { currentLoop->addVertex(x,y); }
void Polygon2D::pushVertex(double x, double y) { currentLoop->pushVertex(x,y); }

void Polygon2D::nextLoop()
{
    currentLoop = new _Polygon2DContour();
    loops.add(currentLoop);
}

void Polygon2D::eraseLastLoop()
{
    if (loops.size() <= 1) return;
    _Polygon2DContour* last = loops[loops.size() - 1];
    for (long int i = 0; i < loops.size(); ) {
        if (loops[i]->getExteriorContour() == last) {
            delete loops[i];
            loops.remove(i);
        }
        else i++;
    }
    if (loops.size() > 1) {
        delete loops[loops.size() - 1];
        loops.remove(loops.size() - 1);
        currentLoop = loops[loops.size() - 1];
    }
}

void Polygon2D::invert()
{
    for (long int i = 0; i < loops.size(); i++) {
        java::ArrayList<Vertex2D>& verts = loops[i]->vertices;
        long int _n = verts.size();
        for (long int _i = 0; _i < _n / 2; _i++) {
            auto _tmp = verts[_i]; verts[_i] = verts[_n - 1 - _i]; verts[_n - 1 - _i] = _tmp;
        }
    }
}

double* Polygon2D::getMinMax()
{
    double* minMax = new double[6];
    double minX = 1e308, minY = 1e308;
    double maxX = -1e308, maxY = -1e308;

    for (long int i = 0; i < loops.size(); i++) {
        for (long int j = 0; j + 1 < loops[i]->vertices.size(); j++) {
            const Vertex2D& v = loops[i]->vertices[j];
            if (v.x > maxX) maxX = v.x;
            if (v.x < minX) minX = v.x;
            if (v.y > maxY) maxY = v.y;
            if (v.y < minY) minY = v.y;
        }
    }

    minMax[0]=minX; minMax[1]=minY; minMax[2]=0;
    minMax[3]=maxX; minMax[4]=maxY; minMax[5]=0;
    return minMax;
}

static bool pointInLoop(const _Polygon2DContour* loop, double x, double y)
{
    bool inside = false;
    java::ArrayList<Vertex2D>& v = const_cast<_Polygon2DContour*>(loop)->vertices;
    long int n = v.size();
    if (n < 3) return false;
    for (long int i = 0, j = n - 1; i < n; j = i++) {
        bool inter = ((v[i].y > y) != (v[j].y > y)) &&
                     (x < (v[j].x - v[i].x) * (y - v[i].y) / ((v[j].y - v[i].y) + 1e-30) + v[i].x);
        if (inter) inside = !inside;
    }
    return inside;
}

Ray* Polygon2D::doIntersection(const Ray& inOut_ray)
{
    RayHit hit;
    if (doIntersection(inOut_ray, &hit) && hit.ray() != nullptr) {
        return new Ray(*hit.ray());
    }
    return nullptr;
}

bool Polygon2D::doIntersection(const Ray& inRay, RayHit* outHit)
{
    if (std::abs(inRay.direction().z()) < VSDK::EPSILON) return false;
    double t = -inRay.origin().z() / inRay.direction().z();
    if (t < 0) return false;

    Vector3Dd p = inRay.origin().add(inRay.direction().multiply(t));
    bool inside = false;
    for (long int i = 0; i < loops.size(); i++) {
        if (pointInLoop(loops[i], p.x(), p.y())) {
            if (loops[i]->getExteriorContour() == nullptr) inside = true;
            else inside = false;
        }
    }
    if (!inside) return false;

    if (outHit != nullptr) {
        if (outHit->shouldStoreRay() || outHit->needsAnySurfaceData()) outHit->setRay(inRay.withT(t));
        else outHit->setHitDistance(t);
        if (outHit->needsPoint()) outHit->p = p;
        if (outHit->needsNormal()) outHit->n = Vector3Dd(0,0,1);
    }
    return true;
}

void Polygon2D::doExtraInformation(const Ray& inRay, double intT, RayHit* outData)
{
    if (outData == nullptr) return;
    outData->p = inRay.origin().add(inRay.direction().multiply(intT));
    outData->n = Vector3Dd(0,0,1);
}

BinaryTreeNode<_Polygon2DContour*>* Polygon2D::getHeadNode() { return headNode; }
void Polygon2D::setHeadNode(BinaryTreeNode<_Polygon2DContour*>* node) { headNode = node; }
